internal/webhooks/patch_validation.go

```diff
@@ -0,0 +1,333 @@
+			if strings.HasPrefix(*jsonPatch.ValueFrom.Variable, "builtin") {
+				if _, ok := builtinVariables[*jsonPatch.ValueFrom.Variable]; !ok {
+					allErrs = append(allErrs,
+						field.Invalid(
+							path.Index(i).Child("valueFrom", "variable"),
+							jsonPatch.ValueFrom.Variable,
+							"not a defined builtin variable",
+						))
+				}
+				continue
+			}
+			if _, ok := variableSet[*jsonPatch.ValueFrom.Variable]; !ok {
+				allErrs = append(allErrs,
+					field.Invalid(
+						path.Index(i).Child("valueFrom", "variable"),
+						jsonPatch.ValueFrom.Variable,
+						fmt.Sprintf("Variable with name %s cannot be found under spec.variables.", *jsonPatch.ValueFrom.Variable),
+					))
+			}
+		}
+	}
+	return allErrs
+}
+
+// This contains a list of all of the valid builtin variables.
+// TODO(killianmuldoon): Match this list to controllers/topology/internal/extensions/patches/variables as those structs become available across the code base i.e. public or top-level internal.
+var builtinVariables = sets.NewString(
+	// Cluster builtins.
+	"builtin.cluster.name",
+	"builtin.cluster.namespace",
+
+	// ClusterTopology builtins.
+	"builtin.cluster.topology.version",
+	"builtin.cluster.topology.class",
+
+	// ControlPlane builtins.
+	"builtin.controlPlane.version",
+	"builtin.controlPlane.replicas",
+
+	// MachineDeployment builtins.
+	"builtin.machineDeployment.version",
+	"builtin.machineDeployment.class",
+	"builtin.machineDeployment.name",
+	"builtin.machineDeployment.topologyName",
+	"builtin.machineDeployment.replicas",
+)
+
+func getClusterClassVariablesMap(clusterClassVariables []clusterv1.ClusterClassVariable) map[string]*clusterv1.ClusterClassVariable {
+	variablesMap := map[string]*clusterv1.ClusterClassVariable{}
+	for i := range clusterClassVariables {
+		variablesMap[clusterClassVariables[i].Name] = &clusterClassVariables[i]
+	}
+	return variablesMap
+}
+
+// Variables with compiled patterns for index checking for array access.
+var (
+	invalidAddIndex     = regexp.MustCompile("^[1-9][0-9]*$")
+	invalidReplaceIndex = regexp.MustCompile("^[0-9][0-9]*$")
+)
+
+// validateIndexAccess checks to see if the jsonPath is attempting to add an element in the array i.e. access by number
+// If the operation is add an error is thrown if a number greater than 0 is used as an index.
+// If the operation is replace an error is thrown if an index is used.
+func validateIndexAccess(jsonPatch clusterv1.JSONPatch, path *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	pathParts := strings.Split(jsonPatch.Path, "/")
+	for _, part := range pathParts {
+		// If the jsonPatch operation is add disallow any number greater than 0. 0 here is a prepend operation on the array.
+		if jsonPatch.Op == "add" && invalidAddIndex.MatchString(part) {
+			allErrs = append(allErrs,
+				field.Invalid(path,
+					jsonPatch.Path,
+					fmt.Sprintf("Arrays can only be accessed using [0] (prepend) or [-] (append). %s not valid", part),
```

Comment:
Please also use lower case here and below
```suggestion
					fmt.Sprintf("arrays can only be accessed using [0] (prepend) or [-] (append)", part),
```
I don't think the last part is required as the field error will print `invalid value: <value>` right in front of the error message. (cf. https://github.com/kubernetes-sigs/cluster-api/pull/5744/files#r759119125)